src/main/java/org/broadinstitute/hellbender/utils/downsampling/ReservoirDownsampler.java

```diff
@@ -116,8 +80,11 @@ public boolean hasFinalizedItems() {
     @Override
     public List<GATKRead> consumeFinalizedItems() {
         if (hasFinalizedItems()) {
-            // pass reservoir by reference rather than make a copy, for speed
-            final List<GATKRead> downsampledItems = reservoir;
+            // one could either:
+            // 1) return copy the reservoir and then clear (but not reallocate) in clearItems, or
+            // 2) return the reservoir by reference and reallocate a new ArrayList in clearItems
+            // 2) is worse because it always reallocates the target sample size, instead of just the reads that are present
+            final List<GATKRead> downsampledItems = new ArrayList<>(reservoir);
```

Comment:
I disagree with your reasoning here. In the previous implementation we: 

1. Returned the existing reservoir by reference (this is O(1))
2. Allocated a new empty reservoir of size targetSampleSize. This is O(number of free heap regions).

In your new implementation we:

1. Allocate a new ArrayList of size reservoir.size(), and copy the existing items from the reservoir into it. This is O(number of free heap regions) for the allocation + O(reservoir.size()) for the copy operation.
2. Call clear() on the existing reservoir. This is O(reservoir.size()) as well!

So I'd argue that the new implementation adds an extra `2 * reservoir.size()` operations that we didn't have before!

Reply:
@droazen You're right, and I will change it.  But why is `clear()` O(size) and not O(1)?  I mean, doesn't it basically amount to `lastUsedIndexInBackingArray = 0` without doing anything to that array?

Reply:
@davidbenjamin If you look at the implementation of `ArrayList.clear()`, you can see that it's a O(n) operation:

```
    public void clear() {
        modCount++;

        // clear to let GC do its work
        for (int i = 0; i < size; i++)
            elementData[i] = null;

        size = 0;
    }
```